sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/JavaOrcColumnarBatchReader.java

```diff
@@ -0,0 +1,503 @@
+        columnVectors[i + partitionIdx].setIsConstant();
+      }
+    }
+
+    // Initialize the missing columns once.
+    for (int i = 0; i < requiredSchema.length(); i++) {
+      if (requestedColIds[i] < 0) {
+        columnVectors[i].putNulls(0, columnarBatch.capacity());
+        columnVectors[i].setIsConstant();
+      }
+    }
+  }
+
+  /**
+   * Return true if there exists more data in the next batch. If exists, prepare the next batch
+   * by copying from ORC VectorizedRowBatch columns to Spark ColumnarBatch columns.
+   */
+  private boolean nextBatch() throws IOException {
+    if (rowsReturned >= totalRowCount) {
+      return false;
+    }
+
+    recordReader.nextBatch(batch);
+    int batchSize = batch.size;
+    if (batchSize == 0) {
+      return false;
+    }
+    rowsReturned += batchSize;
+    for (WritableColumnVector vector : columnVectors) {
+      vector.reset();
+    }
+    columnarBatch.setNumRows(batchSize);
+    int i = 0;
+    while (i < requiredSchema.length()) {
+      StructField field = requiredSchema.fields()[i];
+      WritableColumnVector toColumn = columnVectors[i];
+
+      if (requestedColIds[i] < 0) {
+        toColumn.appendNulls(batchSize);
+      } else {
+        ColumnVector fromColumn = batch.cols[requestedColIds[i]];
+
+        if (fromColumn.isRepeating) {
+          if (fromColumn.isNull[0]) {
+            toColumn.appendNulls(batchSize);
+          } else {
+            DataType type = field.dataType();
+            if (type instanceof BooleanType) {
+              toColumn.appendBooleans(batchSize, ((LongColumnVector)fromColumn).vector[0] == 1);
+            } else if (type instanceof ByteType) {
+              toColumn.appendBytes(batchSize, (byte)((LongColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof ShortType) {
+              toColumn.appendShorts(batchSize, (short)((LongColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof IntegerType || type instanceof DateType) {
+              toColumn.appendInts(batchSize, (int)((LongColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof LongType) {
+              toColumn.appendLongs(batchSize, ((LongColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof TimestampType) {
+              toColumn.appendLongs(batchSize, fromTimestampColumnVector((TimestampColumnVector)fromColumn, 0));
+            } else if (type instanceof FloatType) {
+              toColumn.appendFloats(batchSize, (float)((DoubleColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof DoubleType) {
+              toColumn.appendDoubles(batchSize, ((DoubleColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof StringType || type instanceof BinaryType) {
+              BytesColumnVector data = (BytesColumnVector)fromColumn;
+              int index = 0;
+              while (index < batchSize) {
```

Comment:
use for loop in java. please update all other places.

Reply:
Yep. It's done.